scrape/scrape.go

```diff
@@ -883,7 +883,8 @@ func (c *scrapeCache) iterDone(flushCache bool) {
 	c.seriesPrev, c.seriesCur = c.seriesCur, c.seriesPrev
 
 	// We have to delete every single key in the map.
-	for k := range c.seriesCur {
+	for k, lset := range c.seriesCur {
+		intern.ReleaseLabels(intern.Global, lset)
```

Comment:
I don't think this is right. This is releasing labels for series even before they go stale.

Reply:
I wrote some tests (which I'll push soon). This isn't "wrong," since there's three refs added to the interned string:

1. The series cache ref for a new series 
2. The staleness ref for a new series 
3. The staleness ref for an existing series 

This will release the staleness ref correctly, though it spends a lot of unnecessary CPU time. Instead of doing this, the staleness ref should only be added for a new series and it should be removed when that series goes stale.

Reply:
Result of my benchmarks:

```
name \ time/op    none.txt    orig.txt    old.txt     one-ref.txt
StringInterner-8  2.13s ± 0%  4.23s ± 0%  6.03s ± 0%   2.39s ± 0%

name \ alloc/op   none.txt    orig.txt    old.txt     one-ref.txt
StringInterner-8  396MB ± 0%  421MB ± 0%  397MB ± 0%   418MB ± 0%

name \ allocs/op  none.txt    orig.txt    old.txt     one-ref.txt
StringInterner-8  4.18M ± 0%  4.41M ± 0%  4.19M ± 0%   4.41M ± 0%
```

where:

* `none.txt`: no interning 
* `orig.txt`: behavior of current release 
* `old.txt`: behavior of the original status of this PR
* `one-ref.txt`: fixed PR that only tracks one reference per series and only interns when the series is added. 

The newest approach is the best tradeoff between cpu/memory usage. I'll push a set of commits now.